sql: emit more tracing events from the stats cache

RaduBerinde opened the pull request:
The stats cache has various "slow" paths (where we need to query the
system table). These are currently only logged if verbosity is high.

This change switches to `VEvent` in most cases, so that these are
visible during tracing (including in statement diagnostics bundles).
This will allow us to diagnose slow planning times, e.g. due to the
stats cache getting full.

Release justification: low-risk change to existing functionality, high
potential benefit for debugging issues.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54035)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54035#-:-MGmv2aA0QzU3nRnoCgX:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/stats/stats_cache.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54035#-MGmvYUX4HgmfimSl3f1:-MGmvYUX4HgmfimSl3f2:bz2eiut) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d41695191dc4bf0b0b91a71b6275691e3b4103b/pkg/sql/stats/stats_cache.go#L192)):*
> ```Go
> 		// We are in the process of grabbing stats for this table. Wait until
> 		// that is complete, at which point e.stats will be populated.
> 		log.VEventf(ctx, 1, "waiting for statistics for table %d", tableID)
> ```

So the difference here is that we're always logging somewhere, but the location depends on the verbosity?
___
*[pkg/sql/stats/stats_cache.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54035#-MGmuh9eArRNKdkOpwTi:-MGmuh9eArRNKdkOpwTj:bqz56zk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d41695191dc4bf0b0b91a71b6275691e3b4103b/pkg/sql/stats/stats_cache.go#L309)):*
> ```Go
> 	log.VEventf(ctx, 1, "refreshing statistics for table %d", tableID)
> 	ctx, span := tracing.ForkCtxSpan(ctx, "refresh-table-stats")
> 	defer tracing.FinishSpan(span)
> ```

Will this wait to finish until the goroutine below finishes?


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: emit more tracing events from the stats cache

The stats cache has various "slow" paths (where we need to query the
system table). These are currently only logged if verbosity is high.

This change switches to `VEvent` in most cases, so that these are
visible during tracing (including in statement diagnostics bundles).
This will allow us to diagnose slow planning times, e.g. due to the
stats cache getting full.

Release justification: low-risk change to existing functionality, high
potential benefit for debugging issues.

Release note: None

RaduBerinde committed:
opt: add more tracing events around planning

Add a few more tracing events around the use of the optimizer.

Release note: None

RaduBerinde reviewed (commented):
TFTR! Updated and added a small commit that adds a couple of more events in the planning code.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54035#-:-MGnuO3E5o7fP_aL6GgS:b-cdgpy8)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/stats/stats_cache.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54035#-MGmvYUX4HgmfimSl3f1:-MGnqp2Y6mbPev8kfLma:btro4bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d41695191dc4bf0b0b91a71b6275691e3b4103b/pkg/sql/stats/stats_cache.go#L192)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

So the difference here is that we're always logging somewhere, but the location depends on the verbosity?
</blockquote></details>

VEvent generates a trace event if we are tracing, and also logs it if the verbosity is high. The previous code did not generate a trace event if verbosity was not set.
___
*[pkg/sql/stats/stats_cache.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54035#-MGmuh9eArRNKdkOpwTi:-MGnr2lt6YgCeWWxMcsh:b-xrin66) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d41695191dc4bf0b0b91a71b6275691e3b4103b/pkg/sql/stats/stats_cache.go#L309)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Will this wait to finish until the goroutine below finishes?
</blockquote></details>

Silly me, fixed. Nice catch! This didn't cause any failures because I don't think we ever have a span in the context currently, but I put it in just in case.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54035#-:-MGosNrO2gp54Z4HYV_G:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
